test/integration/apiserver/oidc/oidc_test.go

```diff
@@ -0,0 +1,505 @@
+			"id-token":                  idToken,
+			"idp-issuer-url":            oidcServerURL,
+			"idp-certificate-authority": caFilePath,
+			"refresh-token":             refreshToken,
+		},
+	}
+
+	return cfg
+}
+
+func startTestAPIServerForOIDC(t *testing.T, oidcURL, oidcClientID, oidcCAFilePath string) *kubeapiserverapptesting.TestServer {
+	t.Helper()
+
+	server, err := kubeapiserverapptesting.StartTestServer(
+		t,
+		kubeapiserverapptesting.NewDefaultTestServerOptions(),
+		[]string{
+			fmt.Sprintf("--oidc-issuer-url=%s", oidcURL),
+			fmt.Sprintf("--oidc-client-id=%s", oidcClientID),
+			fmt.Sprintf("--oidc-ca-file=%s", oidcCAFilePath),
+			fmt.Sprintf("--oidc-username-prefix=%s", defaultOIDCUsernamePrefix),
+			fmt.Sprintf("--authorization-mode=%s", modes.ModeRBAC),
+		},
+		framework.SharedEtcd(),
+	)
+	require.NoError(t, err)
+
+	return &server
+}
+
+func fetchOIDCCredentials(t *testing.T, oidcTokenURL string, caCertContent []byte) (idToken, refreshToken string) {
+	t.Helper()
+
+	req, err := http.NewRequest(http.MethodGet, oidcTokenURL, http.NoBody)
+	require.NoError(t, err)
+
+	caPool := x509.NewCertPool()
+	ok := caPool.AppendCertsFromPEM(caCertContent)
+	require.True(t, ok)
+
+	client := http.Client{Transport: &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: caPool,
+		},
+	}}
+
+	token := new(utilsoidc.Token)
+
+	resp, err := client.Do(req)
+	require.NoError(t, err)
+
+	err = json.NewDecoder(resp.Body).Decode(token)
+	require.NoError(t, err)
+
+	return token.IDToken, token.RefreshToken
+}
+
+func fetchExpiredToken(t *testing.T, oidcServer *utilsoidc.TestServer, caCertContent []byte, signingPrivateKey *rsa.PrivateKey) (expiredToken, stubRefreshToken string) {
+	t.Helper()
+
+	tokenURL, err := oidcServer.TokenURL()
+	require.NoError(t, err)
+
+	configureOIDCServerToReturnExpiredIDToken(t, 1, oidcServer, signingPrivateKey)
+	expiredToken, stubRefreshToken = fetchOIDCCredentials(t, tokenURL, caCertContent)
+
+	return expiredToken, stubRefreshToken
+}
+
+func configureOIDCServerToReturnExpiredIDToken(t *testing.T, returningExpiredTokenTimes int, oidcServer *utilsoidc.TestServer, signingPrivateKey *rsa.PrivateKey) {
+	t.Helper()
+
+	oidcServer.TokenHandler().EXPECT().Token().Times(returningExpiredTokenTimes).DoAndReturn(func() (utilsoidc.Token, error) {
+		token, err := utilsoidc.TokenHandlerBehaviourReturningPredefinedJWT(
+			t,
+			signingPrivateKey,
+			oidcServer.URL(),
+			defaultOIDCClientID,
+			defaultOIDCClaimedUsername,
+			defaultStubAccessToken,
+			defaultStubRefreshToken,
+			time.Now().Add(-time.Millisecond).Unix(),
+		)()
+		return token, err
+	})
+}
+
+func configureOIDCServerToReturnExpiredRefreshTokenErrorOnTryingToUpdateIDToken(oidcServer *utilsoidc.TestServer) {
+	oidcServer.TokenHandler().EXPECT().Token().Times(2).Return(utilsoidc.Token{}, utilsoidc.ErrRefreshTokenExpired)
+}
+
+func generateCert(t *testing.T) (cert, key []byte, certFilePath, keyFilePath string) {
```

Comment:
Use `t.TempDir` instead of `os.TempDir` so all cleanup is handled by the test logic.  Also, use a single tmp dir in this function, it currently creates three.